ee/query-service/license/manager.go

```diff
@@ -0,0 +1,287 @@
+func (lm *Manager) LoadActiveLicense() error {
+	var err error
+	active, err := lm.repo.GetActiveLicense(context.Background())
+	if err != nil {
+		return err
+	}
+	if active != nil {
+		lm.SetActive(active)
+	} else {
+		zap.S().Info("No active license found.")
+	}
+
+	return nil
+}
+
+func (lm *Manager) GetLicenses(ctx context.Context) (response []*model.License, apiError *model.ApiError) {
+
+	licenses, err := lm.repo.GetLicenses(ctx)
+	if err != nil {
+		return nil, model.InternalError(err)
+	}
+
+	for _, l := range licenses {
+		l.ParsePlan()
+
+		if l.Key == lm.activeLicense.Key {
+			l.IsCurrent = true
+		}
+
+		if l.ValidUntil == -1 {
+			l.ValidUntil = l.ValidFrom + 31556926
+		}
+
+		response = append(response, &l)
+	}
+	return
+}
+
+// Validator validates license after an epoch of time
+func (lm *Manager) Validator(ctx context.Context) {
+	defer close(lm.terminated)
+	tick := time.NewTicker(validationFrequency)
+	defer tick.Stop()
+
+	lm.Validate(ctx)
+
+	for {
+		select {
+		case <-lm.done:
+			return
+		default:
+			select {
+			case <-lm.done:
+				return
+			case <-tick.C:
+				lm.Validate(ctx)
+			}
+		}
+
+	}
+}
+
+// Validate validates the current active license
+func (lm *Manager) Validate(ctx context.Context) (reterr error) {
+	zap.S().Info("License validation started")
+	if lm.activeLicense == nil {
+		return nil
+	}
+
+	defer func() {
+		lm.mutex.Lock()
+
+		lm.lastValidated = time.Now().Unix()
+		if reterr != nil {
+			zap.S().Errorf("License validation completed with error", reterr)
+			atomic.AddUint64(&lm.failedAttempts, 1)
+			telemetry.GetInstance().SendEvent(telemetry.TELEMETRY_LICENSE_CHECK_FAILED,
+				map[string]interface{}{"err": reterr.Error()})
+		} else {
+			zap.S().Info("License validation completed with no errors")
+		}
+
+		lm.mutex.Unlock()
+	}()
+
+	response, apiError := validate.ValidateLicense(lm.activeLicense.ActivationId)
+	if apiError != nil {
+		zap.S().Errorf("failed to validate license", apiError)
+		return apiError.Err
+	}
+
+	if response.PlanDetails == lm.activeLicense.PlanDetails {
+		// license plan hasnt changed, nothing to do
+		return nil
+	}
+
+	if response.PlanDetails != "" {
+
+		// copy and replace the active license record
+		l := model.License{
+			Key:               lm.activeLicense.Key,
+			CreatedAt:         lm.activeLicense.CreatedAt,
+			PlanDetails:       response.PlanDetails,
+			ValidationMessage: lm.activeLicense.ValidationMessage,
+			ActivationId:      lm.activeLicense.ActivationId,
+		}
+
+		if err := l.ParsePlan(); err != nil {
+			zap.S().Errorf("failed to parse updated license", zap.Error(err))
+			return err
+		}
+
+		// updated plan is parsable, check if plan has changed
+		if lm.activeLicense.PlanDetails != response.PlanDetails {
+			err := lm.repo.UpdatePlanDetails(ctx, lm.activeLicense.Key, response.PlanDetails)
+			if err != nil {
+				// unexpected db write issue but we can let the user continue
+				// and wait for update to work in next cycle.
```

Comment:
yeh but it will get corrected in next restart.  The problem is the error capturing here is cyclical, when do you stop. So in a rare event that the db write fails (which is very unlikely), the next restart cycle - validate() will be called and the db will be corrected. 

The transaction here is eventually consistent.